routes/[version]/apikey.js

```diff
@@ -0,0 +1,55 @@
+const { checkAuth } = require('../../middlewares/auth');
+const { selectAPIToken, setAPIToken, removeAPIToken } = require('../../services/db');
+
+module.exports = {
+    get: [checkAuth, get],
+    put: [checkAuth, put],
+    delete: [checkAuth, del],
+};
+
+function generateAPIToken() {
+    let result = '',
+        counter = 0;
+    const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789_';
+    const charlen = chars.length;
+
+    while (counter < 11) {
+        result += chars.charAt(Math.floor(Math.random() * charlen));
+        counter += 1;
+    }
+    return 'METIS' + result;
+}
+
+/**
+ * @api {put} /v0/apikey Create new API key
+ * @apiName SetAPIKey
+ * @apiGroup Users
+ * @apiDescription Create new API key
+ * @apiPermission GUI_ONLY
+ */
+async function put(req, res, next) {
+    return res.json({ apikey: await setAPIToken(req.user.id, generateAPIToken()) });
+}
+
+/**
+ * @api {del} /v0/apikey Remove API key
+ * @apiName RemoveAPIKey
+ * @apiGroup Users
+ * @apiDescription Remove all API keys
+ * @apiPermission GUI_ONLY
+ */
+async function del(req, res, next) {
+    removeAPIToken(req.user.id);
+    return res.json({});
+}
+
+/**
+ * @api {get} /v0/apikey Check if API key exists
+ * @apiName GetAPIKey
+ * @apiGroup Users
+ * @apiDescription Get API key, if any
+ * @apiPermission GUI_ONLY
+ */
+async function get(req, res, next) {
+    return res.json({ apikey: await selectAPIToken(req.user.id) });
```

Comment:
I think this is insecure. I like to return nothing with response code 204 if set and nothing with response code 404 if not set.

Reply:
Why insecure? This endpoint requires authorization.

Reply:
API key is not scoped (=full control). API key doesn't have any expiration date. Also, if it is possible to show the token some time after creation, then everyone will immediately realize that the tokens are stored in the database unencrypted. Such a shame.
This product doesn't seem to require any particularly high security requirements, but just not giving the token away unnecessarily isn't difficult either.

Reply:
Created #88 to keep up with this in future (we don't have time for this now)